test: Makes `wait_for_getdata` delete data on checks, plus allows to check the getdata message type

sr-gi opened the pull request:
https://github.com/bitcoin/bitcoin/pull/18690 introduced a way of checking if a collection of hashes was present in the last received getdata message. However, bits of manual checking were left in some tests.

This changes the behavior of `wait_for_getdata` so data is deleted on checks, plus allows checking for the `getdata` inv types so we don't have to do it manually (when needed)

<!--
*** Please remove the following help text before submitting: ***

Pull requests without a rationale and clear improvement may be closed
immediately.

GUI-related pull requests should be opened against
https://github.com/bitcoin-core/gui
first. See CONTRIBUTING.md
-->

<!--
Please provide clear motivation for your patch and explain how it improves
Bitcoin Core user experience or Bitcoin Core developer experience
significantly:

* Any test improvements or new tests that improve coverage are always welcome.
* All other changes should have accompanying unit tests (see `src/test/`) or
  functional tests (see `test/`). Contributors should note which tests cover
  modified code. If no tests exist for a region of modified code, new tests
  should accompany the change.
* Bug fixes are most welcome when they come with steps to reproduce or an
  explanation of the potential issue as well as reasoning for the way the bug
  was fixed.
* Features are welcome, but might be rejected due to design or scope issues.
  If a feature is based on a lot of dependencies, contributors should first
  consider building the system outside of Bitcoin Core, if possible.
* Refactoring changes are only accepted if they are required for a feature or
  bug fix or otherwise improve developer experience significantly. For example,
  most "code style" refactoring changes require a thorough explanation why they
  are useful, what downsides they have and why they *significantly* improve
  developer experience or avoid serious programming bugs. Note that code style
  is often a subjective matter. Unless they are explicitly mentioned to be
  preferred in the [developer notes](/doc/developer-notes.md), stylistic code
  changes are usually rejected.
-->

<!--
Bitcoin Core has a thorough review process and even the most trivial change
needs to pass a lot of eyes and requires non-zero or even substantial time
effort to review. There is a huge lack of active reviewers on the project, so
patches often sit for a long time.
-->

sr-gi committed:
test: Cleans some manual checks/drops when using wait_for_getdata

https://github.com/bitcoin/bitcoin/pull/18690 introduced a way of checking
if a collection of hashes was present in the last received getdata message.
However, bits of manual checking were left in some tests.

AngusP reviewed (approved):
ACK 256dc9745241b247262cc69beaf9bc57786b4d55 (caveat I'm pretty new to this code)

AngusP reviewed (approved):
ACK df90ea6c73412736608ffc6074b73da980ad25a5

naiyoma commented:
Concept ACK, 
nit: Could you also take a look at `announce_tx_and_wait_for_getdata` in `p2p_segwit.py` for similar cleanup?

sr-gi commented:
> Concept ACK, nit: Could you also take a look at `announce_tx_and_wait_for_getdata` in `p2p_segwit.py` for a similar cleanup?

Unfortunately, that's not the case given how the tests using `announce_tx_and_wait_for_getdata` are built. That is, the same transaction may be sent multiple times using `announce_tx_and_wait_for_getdata`, so you need to potentially pop the old one to make sure the assert is not using old data.

The only way of working around this would be making `wait_for_getdata` pop the data instead of just reading it (similarly to how #29736 is approached) but that may need it's own PR

sr-gi committed:
test: makes wait_for_getdata pops data on checks, plus check the message types if provided

Currently, `wait_for_getdata` checks data in `last_message`, but it leaves is as is. This results
in multiple tests having to manually pop data from the collection is multiple getdata messages need
to be checked in order to make sure that we are not asserting the oldest message. This was partially
solved by checking the hash of what was being waited for, but the issue remains if the same tx/block
is sent multiple times during the test.

In some of this scenarios, we were also manually checking the getdata message type manually which is not
possible anymore if we pop data on checking, so `wait_for_getdata` is extended to support checking the type
when we care about it

sr-gi commented:
> Concept ACK, nit: Could you also take a look at `announce_tx_and_wait_for_getdata` in `p2p_segwit.py` for a similar cleanup?

Covered that in [067b009](https://github.com/bitcoin/bitcoin/pull/29748/commits/067b009bbf47f7bc5adeb6b500042f7c44bfb03f) by deleting the data on checks (pop instead of get) plus allowing to check the inv type.

For the latter, I've done it in a way that all items need to be of the same type. We could address this so every single inv has its own type, but for that we would be required to provide a list of invs to `wait_for_getdata`, instead of a list of hashes. If this is something we want to have, it would certainly need its own PR given it'd require a way bigger change.

AngusP reviewed (approved):
reACK 067b009bbf47f7bc5adeb6b500042f7c44bfb03f

rkrux reviewed (approved):
tACK [067b009](https://github.com/bitcoin/bitcoin/pull/29748/commits/067b009bbf47f7bc5adeb6b500042f7c44bfb03f)

Make successful, functional tests successful

I believe this is a good refactor introducing `check_last_getdata` and modifying `wait_for_getdata` that will reduce some level of code duplication within tests.

Regarding this comment, I tried this refactor and the tests pass.
> check_last_getdata uses the exact same logic as the internal function for wait_for_getdata. It feels like there should be a way of defining it as a lambda function, and getting rid of the internal function. I haven't been able to figure out how though.

```diff
diff --git a/test/functional/test_framework/p2p.py b/test/functional/test_framework/p2p.py
index 684f029d58..a28b8860e4 100755
--- a/test/functional/test_framework/p2p.py
+++ b/test/functional/test_framework/p2p.py
@@ -430,6 +430,15 @@ class P2PConnection(asyncio.Protocol):
         logger.debug(log_message)


+def assert_last_data(last_data, hash_list, datatype):
+    if not last_data:
+        return False
+    if datatype is None:
+        return [x.hash for x in last_data.inv] == hash_list
+    else:
+        return [(x.hash, x.type) for x in last_data.inv] == [(h, datatype) for h in hash_list]
+
+
 class P2PInterface(P2PConnection):
     """A high-level P2P interface class for communicating with a Bitcoin node.

@@ -603,9 +612,7 @@ class P2PInterface(P2PConnection):
     def check_last_getdata(self, hash_list):
         with p2p_lock:
             last_data = self.last_message.get("getdata")
-            if last_data is None:
-                return False
-            return [x.hash for x in last_data.inv] == hash_list
+            return assert_last_data(last_data, hash_list, None)

     def wait_for_tx(self, txid, *, timeout=60):
         def test_function():
@@ -645,12 +652,7 @@ class P2PInterface(P2PConnection):
         The object hashes in the inventory vector must match the provided hash_list."""
         def test_function():
             last_data = self.last_message.pop("getdata", None)
-            if not last_data:
-                return False
-            if datatype is None:
-                return [x.hash for x in last_data.inv] == hash_list
-            else:
-                return [(x.hash, x.type) for x in last_data.inv] == [(h, datatype) for h in hash_list]
+            return assert_last_data(last_data, hash_list, datatype)
```

sr-gi commented:
> Regarding this comment, I tried this refactor and the tests pass.

Thanks! I think I may have been locking the `p2p_lock` twice which made test timeout.

rkrux commented on test/functional/test_framework/p2p.py:
```diff
@@ -600,6 +600,19 @@ def test_function():
     # Message receiving helper methods
 
+    def _check_last_getdata(self, hash_list, datatype=None):
+        last_data = self.last_message.pop("getdata", None)
```
Note: This will pop from last message via `check_last_getdata()` call as well, which is different from previous commits. That's why I accepted `last_data` as an argument in my suggestion.
Is popping in `check_last_getdata` also fine?

sr-gi replied:
Yeah, I noticed that on your suggestion, but I don't think we do reuse the value in any test (same as with `wait_for_*`)

sr-gi commented:
Rebased to fixed merge conflicts after #29736

AngusP commented:
tACK e233ef18364a5758adcb83eea53bdf92aa1bf551
